---
title: "P1 | Basic tools for data visualization"
subtitle: "Introduction"
author: "<b>Marta Coronado Zamora and Miriam Merenciano</b>"
date: "19 February 2026"
output:
  xaringan::moon_reader:
    css: [xaringan-themer.css, rutgers, "https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.2/animate.min.css", "https://use.fontawesome.com/releases/v7.0.0/css/all.css"]    
    lib_dir: libs
    self_contained: true
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      ratio: "16:9"
      
---
```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
# Animations: https://github.com/daneden/animate.css#animations
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, fig.width = 4, fig.height = 3.5, fig.show = 'hold', message = FALSE, warning = F,  fig.retina = 3)
library(ggplot2)
library(lattice)
library(tidyverse)
# pagedown::chrome_print("/Users/miriam/Desktop/DATA-VISUALIZATION/P1-slides.html")
```

class: center, animated, bounceInDown

<style>
.title-slide {
  background-image: url(img/upc.png);
  background-size: 300px;
}
</style>

# Keep in touch

| Miriam Merenciano |
|:-:|
| <a href="mailto:miriam.merenciano@univ-lyon1.fr"><i class="fa fa-paper-plane fa-fw"></i> miriam.merenciano@univ-lyon1.fr </a> | 
|  <a href="https://lbbe.univ-lyon1.fr/fr/"><i class="fa fa-map-marker fa-fw"></i>&nbsp; Laboratory of Biometry and Evolutionary Biology </a> |

---
layout: true
class: animated, fadeIn

---
# Practical session dynamics

**Content** (P1-P5)
- Introduction 
- Exercises - complete and submit to [ATENEA](https://atenea.upc.edu/login/index.php)
- Project (divided in to 2 assignments)
<br><br>

--

**Interactive <i class="fab fa-r-project fa-fw"></i></i> documents**  
`R code` can be executed within RStudio!

```{r, echo=TRUE}
value <- 2
value + 3
```

---
layout: false
class: left, bottom, inverse, animated, bounceInDown

# Get started!
## **Tools for data visualization**

---
layout: true
class: animated, fadeIn

---

# Practice <i class="fas fa-cogs"></i>
## Introduction to `ggplot2`

- Open the document `P1_exercises.Rmd` in RStudio and complete the exercises.
- Upload the completed document to [ATENEA](https://atenea.upc.edu/login/index.php) at the end of the session.

---

# Project

## Group project

The project has 3 different parts (A, B and C) divided in two big assignments.
- You can deliver the parts separately to get feedback before submitting the final version
- Each part must be submitted before next practical session
- The first assignment will contain parts A and B
- The second assignment will contain part C
- ~15 minutes in the end of each class devoted to discuss your problems

---
# Project

## Group project

__Part A__

- __1\.__ Create groups of \~4 people
- __2\.__ Choose a data set with the following requirements
    + Tabular format (txt, csv, tsv...)
    + More than 80 observations
    + At least 6 variables
    + At least 2 discrete and 3 continuous variables
    + Data with biological meaning
    + Different from the ones chosen by other groups

---
# Project

## Group project

- __3\.__ Describe your data set:
    + Where and why was the information collected?
    + Which is the meaning of each variable?
    + Do the variables have unit? Which one?
    + Does the data set have a long format?
- __4\.__ Write the code to:
    + Read it into R
    + Reshape the data if necessary into long format
    + Check the variable classes and update them if necessary

Write 3 and 4 in an `R Markdown` document and __submit it before next practical session__ (one per group).

<i class="fas fa-info-circle"></i> If you need help formatting the R Markdown, ask me for a guide of an introduction to R Markdown.

---

## Data sets from research articles

- ["Zika virus evolution and spread in the Americas"](https://www.nature.com/articles/nature22402#supplementary-information) (Table S2)
- ["Great ape genetic diversity and population history"](https://www.nature.com/articles/nature12228#supplementary-information) (Table S1 or S3)
- ["Transcriptome and genome sequencing uncovers functional variation in humans"](https://www.nature.com/articles/nature12531). [Table with cis eQTLs in EUR](https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/EUR373.gene.cis.FDR5.best.rs137.txt.gz) ([description](https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/GeuvadisRNASeqAnalysisFiles_README.txt))
- ["Signatures of archaic adaptive introgression in present-day human Populations"](https://academic.oup.com/mbe/article/34/2/296/2633371#supplementary-data) (Table S3)
- ["The evolutionary history of dogs in the Americas"](http://science.sciencemag.org/content/361/6397/81) (Table S1)
- ["Ancient genomes document multiple waves of migration in Southeast Asian prehistory"](http://science.sciencemag.org/content/361/6397/92) (Table S1)
- ["Population-scale long-read sequencing uncovers transposable elements associated with gene expression variation and adaptive signatures in _Drosophila_"](https://www.nature.com/articles/s41467-022-29518-8#Sec29) (Table S10)
- ["Comprehensive characterization of 536 patient-derived xenograft models prioritizes candidates for targeted treatment"](https://www.nature.com/articles/s41467-021-25177-3#Sec31) (Table S1)
- ["Pan-cancer analysis of whole genomes"](https://www.nature.com/articles/s41586-020-1969-6#Sec23) (Table S1)
- ["The genomic basis of copper tolerance in _Drosophila_ is shaped by a complex interplay of regulatory and environmental factors"](https://static-content.springer.com/esm/art%3A10.1186%2Fs12915-022-01479-w/MediaObjects/12915_2022_1479_MOESM4_ESM.xlsx) (Table S3)
- ["Transposons contribute to the diversification of the head, gut, and ovary transcriptomes across _Drosophila_ natural strains"](https://raw.githubusercontent.com/GonzalezLab/chimerics-transcripts-dmelanogaster/main/01.%20Chimeric%20gene-TE%20transcripts/chimerics_v2.tab) (Chimeric gene-TE transcripts data)

